src/main/scala/com/softwaremill/realworld/articles/comments/CommentsRepository.scala

```diff
@@ -52,56 +54,74 @@ class CommentsRepository(quill: Quill.Sqlite[SnakeCase]):
+        ar <- queryArticle.join(ar => cr.articleId == ar.articleId)
+      } yield (pr.userId, ar.articleId)
     )
-      .map(_.flatten.headOption)
+      .map(_.headOption)
 
-  def findComment(commentId: Int): Task[Option[Comment]] =
-    run(
-      for {
-        cr <- queryComment if cr.commentId == lift(commentId)
-        ar <- queryProfile.leftJoin(ar => ar.userId == cr.authorId)
-      } yield (cr, ar)
-    )
+  def findComment(commentId: Int, viewerId: Int): Task[Option[Comment]] = {
+    val commentRow: Quoted[Query[CommentRow]] = quote {
+      queryComment.filter(cr => cr.commentId == lift(commentId))
+    }
+
+    val commentQuery = buildCommentQueryWithFollowing(commentRow, viewerId)
+
+    run(commentQuery)
       .map(_.headOption)
-      .map(_.flatMap(comment))
+      .map(_.map(comment))
+  }
 
-  def findComments(articleId: Int): Task[List[Comment]] =
-    run(
-      for {
-        cr <- queryComment if cr.articleId == lift(articleId)
-        ar <- queryProfile.leftJoin(ar => ar.userId == cr.authorId)
-      } yield (cr, ar)
-    )
-      .map(_.flatMap(comment))
-
-  private def comment(tuple: (CommentRow, Option[ProfileRow])): Option[Comment] = {
-    val (cr, arOpt) = tuple
-
-    arOpt.map(ar =>
-      Comment(
-        id = cr.commentId,
-        createdAt = cr.createdAt,
-        updatedAt = cr.updatedAt,
-        body = cr.body,
-        author = CommentAuthor(
-          username = ar.username,
-          bio = ar.bio,
-          image = ar.image,
-          // TODO implement "following"
-          following = false
-        )
+  def findComments(articleId: Int, viewerIdOpt: Option[Int]): Task[List[Comment]] = {
+    val commentRow: Quoted[Query[CommentRow]] = quote {
+      queryComment.filter(cr => cr.articleId == lift(articleId))
+    }
+
+    val commentQuery = viewerIdOpt match
+      case Some(viewerId) => buildCommentQueryWithFollowing(commentRow, viewerId)
+      case None           => buildCommentQuery(commentRow)
+
+    run(commentQuery).map(_.map(comment))
+  }
+
+  private def comment(tuple: (CommentRow, ProfileRow, Boolean)): Comment = {
+    val (cr, pr, isFollowing) = tuple
+
+    Comment(
+      id = cr.commentId,
+      createdAt = cr.createdAt,
+      updatedAt = cr.updatedAt,
+      body = cr.body,
+      author = CommentAuthor(
+        username = pr.username,
+        bio = pr.bio,
+        image = pr.image,
+        following = isFollowing
       )
     )
   }
 
+  private def buildCommentQuery(crq: Quoted[Query[CommentRow]]) = {
+    quote {
+      for {
+        cr <- crq
+        pr <- queryProfile.join(ar => ar.userId == cr.authorId)
+      } yield (cr, pr, false)
+    }
+  }
+
+  private def buildCommentQueryWithFollowing(crq: Quoted[Query[CommentRow]], viewerId: Int) = {
+    quote {
+      for {
+        tuple <- buildCommentQuery(crq)
+        isFollowing = queryFollower
+          .filter(f => (f.userId == tuple._2.userId) && (f.followerId == lift(viewerId)))
+          .map(_ => 1)
```

Comment:
**question:** Why map to 1, if you actually only care about `.nonEmpty`?

Reply:
Without .map(_ => 1), there are many errors during tests like these: 

```
      Exception in thread "zio-fiber-4132" java.lang.IllegalStateException: 
      When synthesizing Joins, Quill found some variables that could not be traced back to their
      origin: List(u, f1). Typically this happens when there are some flatMapped
      clauses that are missing data once they are flattened.
      Sometimes this is the result of a internal error in Quill. If that is the case, please
      reach out on our discord channel https://discord.gg/2ccFBr4 and/or file an issue
      on https://github.com/zio/zio-quill.
      Faulty expression: '`querySchema`("users").join(`querySchema`("favorites_articles")).on((u, f1) => ((u.email == lift(72ede273-2994-4ac1-aef8-eaf6dbc38437)) && (f1.articleId == ar.articleId)) && (f1.profileId == u.userId)).map(uf1 => (u, f1)).nonEmpty'. Free variables: 'List(u, f1)'.
 ```

Reply:
Interesting, my bad for interpreting these queries as ordinary types. Forget my comments about `map` and `exist` then ;)